presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,268 @@
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.prestosql.execution.StageState.FINISHED;
+import static io.prestosql.execution.StageState.RUNNING;
+import static io.prestosql.execution.StageState.SCHEDULED;
+import static io.prestosql.execution.StageState.SCHEDULING_SPLITS;
+import static io.prestosql.spi.predicate.TupleDomain.columnWiseUnion;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final ConcurrentHashMap<SourceDescriptor, SettableFuture<TupleDomain<String>>> dynamicFilterSummaries = new ConcurrentHashMap<>();
+    private final ConcurrentMap<QueryId, Supplier<List<StageInfo>>> queries = new ConcurrentHashMap<>();
+    private final ScheduledExecutorService collectDynamicFiltersExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("DynamicFilterService"));
+
+    @Inject
+    public DynamicFilterService(TaskManagerConfig taskConfig)
+    {
+        collectDynamicFiltersExecutor.scheduleWithFixedDelay(this::collectDynamicFilters, 0, taskConfig.getStatusRefreshMaxWait().toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    public void registerQuery(QueryId queryId, Supplier<List<StageInfo>> stageInfoSupplier)
+    {
+        queries.putIfAbsent(queryId, stageInfoSupplier);
+    }
+
+    @VisibleForTesting
+    public void collectDynamicFilters()
+    {
+        List<QueryId> queriesWithDynamicFilters = dynamicFilterSummaries.entrySet()
+                .stream()
+                .filter(entry -> !entry.getValue().isDone())
+                .map(entry -> entry.getKey().getQueryId())
+                .distinct()
+                .collect(toImmutableList());
+
+        for (QueryId queryId : queriesWithDynamicFilters) {
+            Supplier<List<StageInfo>> stageInfoSupplier = queries.get(queryId);
+            if (stageInfoSupplier != null) {
```

Comment:
I don't think `stageInfoSupplier` can return null. Use `requireNotNull` instead

Reply:
`getSummary` gets called through `DistributedExecutionPlanner#plan` whereas `registerQuery` is called asynchronously on state change of query to STARTING state. So I thought it's possible for `dynamicFilterSummaries` to have a QueryId which hasn't yet been added to `queries` structure.